OMCompiler/Compiler/SimCode/SimCodeUtil.mo

```diff
@@ -9100,6 +9107,58 @@ algorithm
   tpl := (index, fmi_index);
 end setVariableIndexHelper2;
 
+protected function getFmiUnitDefinitions
+  "returs the list<UnitDefinitions> which needs to exported in modelDescription.xml"
+  input array<list<SimCodeVar.SimVar>> simVars;
+  output list<SimCode.UnitDefinition> unitDefinitions = {};
+protected
+  HashSetString.HashSet unitNameKeys;
+algorithm
+  //special order for fmi: real => intger => boolean => string => external
+  unitNameKeys := HashSetString.emptyHashSet();
+  for i in SimVarsIndex.state : SimVarsIndex.stringConst loop
+    (unitDefinitions, unitNameKeys) := getFmiUnitDefinitionsHelper(Dangerous.arrayGetNoBoundsChecking(simVars, Integer(i)), unitDefinitions, unitNameKeys);
+  end for;
+  //print("\n Final Units List :" + anyString(unitDefinitions));
+end getFmiUnitDefinitions;
+
+protected function getFmiUnitDefinitionsHelper
+  "helper function which creates the list<UnitDefintions> to be exported in modelDescription.xml"
+  input list<SimCodeVar.SimVar> inVars;
+  input output list<SimCode.UnitDefinition> unitDefinitions;
+  input output HashSetString.HashSet unitNameKeys;
+protected
+  Unit.Unit unit;
+  SimCode.BaseUnit baseUnit;
+algorithm
+  for var in inVars loop
+    if isSome(var.exportVar) then
+      // check if the var has unit and also check if the unit is already added to hashset
+      if not stringEq(var.unit, "") and not BaseHashSet.has(var.unit, unitNameKeys) then
+        unitNameKeys := BaseHashSet.add(var.unit, unitNameKeys);
+        try
+          unit := Unit.parseUnitString(var.unit); // get the SI- units information
+          unitDefinitions := SimCode.UNITDEFINITION(var.unit, transformUnitToBaseUnit(unit)) :: unitDefinitions;
+        else
+          // catch the units which are not calculated
+        end try;
+      end if;
+    end if;
+  end for;
+end getFmiUnitDefinitionsHelper;
+
+public function transformUnitToBaseUnit
+  "translate Unit.UNIT to SimCode.BASEUNIT"
+  input Unit.Unit unit;
+  output SimCode.BaseUnit baseUnit;
+protected
+  Integer mol, cd, m, s, A, K, kg;
+  Real factors;
+algorithm
+  Unit.UNIT(factors, mol, cd, m, s, A, K, kg) := unit;
+  baseUnit := SimCode.BASEUNIT(mol, cd, m, s, A, K, kg, factors, 0.0);
```

Comment:
The shift should cause problems for some units, e.g. degree Fahrenheit.

Reply:
So should we remove the shift ?

Reply:
I know; the unit module doesn't support shifts right now. But that means we will get wrong results if there is actually a shift in the unit. We probably have to fix it directly in the unit module.

Reply:
And btw: There is no such unit in the tests we are using right now. So we should also add a new test case to test units specifically.